### PR TITLE
Refactor RenderState creation

### PR DESCRIPTION
**Pull request by lilleyse**
For #640

### CODE REVIEW

**Comment by lilleyse** [comment]
Not sure about the documentation in RenderState.js. What's the general policy for documentation in internal files?

File: Source/Renderer/ClearCommand.js [hunk 1]
```diff
@@ -53,7 +53,7 @@ define([
          *
          * @default undefined
          *
-         * @see Context#createRenderState
+         * @see RenderState
```
Comment (pjcozzi) [inline Source/Renderer/ClearCommand.js]: Remove this.  Note the `@type` above.

File: Source/Renderer/DrawCommand.js [hunk 2]
```diff
@@ -117,7 +117,7 @@ define([
          * @type {RenderState}
          * @default undefined
          *
-         * @see Context#createRenderState
+         * @see RenderState
```
Comment (pjcozzi) [inline Source/Renderer/DrawCommand.js]: Same comment.

File: Source/Renderer/RenderState.js [hunk 3]
```diff
@@ -173,12 +182,12 @@ define([
             (!defined(viewport.width)) ? context.drawingBufferWidth : viewport.width,
             (!defined(viewport.height)) ? context.drawingBufferHeight : viewport.height) : undefined;
 
+        //>>includeStart('debug', pragmas.debug);
```
Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Below is a `RuntimeError` so we do not want to optimize it out in release builds.

Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Actually, I would say this `RuntimeError` and the one below could really be `DeveloperError`s.

File: Source/Renderer/RenderState.js [hunk 4]
```diff
@@ -272,12 +278,179 @@ define([
...
+    var renderStateCache = {};
+
+    /**
+     * Validates and then finds or creates an immutable render state, which defines the pipeline
+     * state for a {@link DrawCommand} or {@link ClearCommand}.  All inputs states are optional.  Omitted states
+     * use the defaults shown in the example below.
+     *
+     * @param {Object} options Object with the following properties:
+     * @param [Object} options.context The context.
+     * @param {Object} options.renderState The states defining the render state as shown in the example below.
```
Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: This is optional so put it in brackets: `[options.renderState]`.

Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Poke around here to learn more about our reference doc: http://usejsdoc.org/index.html

File: Source/Renderer/RenderState.js [hunk 5]
```diff
@@ -272,12 +278,179 @@ define([
...
+     * @exception {DeveloperError} Invalid renderState.stencilTest.frontOperation.zPass.
+     * @exception {DeveloperError} Invalid renderState.stencilTest.backOperation.fail.
+     * @exception {DeveloperError} Invalid renderState.stencilTest.backOperation.zFail.
+     * @exception {DeveloperError} Invalid renderState.stencilTest.backOperation.zPass.
+     * @exception {DeveloperError} renderState.viewport.width must be greater than or equal to zero.
+     * @exception {DeveloperError} renderState.viewport.width must be less than or equal to the maximum viewport width.
+     * @exception {DeveloperError} renderState.viewport.height must be greater than or equal to zero.
+     * @exception {DeveloperError} renderState.viewport.height must be less than or equal to the maximum viewport height.
+     *
+     * @see DrawCommand
```
Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Remove these `@see`s.

File: Source/Renderer/RenderState.js [hunk 6]
```diff
@@ -272,12 +278,179 @@ define([
...
+    RenderState.fromCache = function(options) {
+        options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+
+        //>>includeStart('debug', pragmas.debug);
+        if (!defined(options.context)) {
+            throw new DeveloperError('options.context is required.');
+        }
+        //>>includeEnd('debug');
+
+        var context = options.context;
```
Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: `context` is only used in one place - there's no need to create the local, just use
```
context : options.context
```
below.

Or move this declaration above and use just
```
if (!defined(context)) {
```

Comment (pjcozzi) [inline Source/Renderer/RenderState.js]: Actually, sorry, I didn't catch this sooner, but I bet we can do this without requiring the context at all, e.g., see:

https://github.com/AnalyticalGraphicsInc/cesium/compare/3d-tiles#diff-3554c38266bd8281183f28deee64d1b2R264

Let's discuss in person.

**Comment by pjcozzi** [comment]
That's all my comments.

**Comment by lilleyse** [comment]
Updated with ContextLimits

**Comment by pjcozzi** [comment]
Can you close and re-open this into the new [renderer-refactor](https://github.com/AnalyticalGraphicsInc/cesium/tree/renderer-refactor) branch?

**Pull request closed without merging**